src/Sylius/Component/Promotion/Generator/CouponGenerator.php

```diff
@@ -72,53 +67,54 @@ public function __construct(
      */
     public function generate(PromotionInterface $promotion, InstructionInterface $instruction)
     {
-        $this->generatedCoupons = [];
+        $generatedCoupons = [];
 
         $this->assertGenerationIsPossible($instruction);
         for ($i = 0, $amount = $instruction->getAmount(); $i < $amount; ++$i) {
-            $code = $this->generateUniqueCode($instruction->getCodeLength());
+            $code = $this->generateUniqueCode($instruction->getCodeLength(), $generatedCoupons);
             $coupon = $this->couponFactory->createNew();
             $coupon->setPromotion($promotion);
             $coupon->setCode($code);
             $coupon->setUsageLimit($instruction->getUsageLimit());
             $coupon->setExpiresAt($instruction->getExpiresAt());
 
-            $this->generatedCoupons[$code] = true;
+            $generatedCoupons[$code] = $coupon;
 
             $this->objectManager->persist($coupon);
         }
 
         $this->objectManager->flush();
 
-        return $this->generatedCoupons;
+        return $generatedCoupons;
     }
 
-
     /**
      * @param int $codeLength
+     * @param array $reservedCoupons
      *
      * @return string
      */
-    protected function generateUniqueCode($codeLength)
+    protected function generateUniqueCode($codeLength, array $reservedCoupons)
```

Comment:
Why not `generatedCoupons`? Using the same array under different name within the same class introduces confusion.